text/3735-field-projections.md

```diff
@@ -0,0 +1,1555 @@
+
+Should we allow user implementations of the `Field` trait and have user-visible internals for it, or
+should we make it more opaque and sealed to reserve the possibility of supporting enums or similar?
+
+## Generalized enum projection
+
+Is there a generalization of enum projection that allows for runtime-conditional projection, for
+structures that sometimes-but-not-always have a given field? This could guarantee the type and
+identity of the field, but require a projection to be validated against a runtime instance of the
+value before confirming that the field exists and providing the projected field type. This would
+allow enums, as well as runtime equivalents such as C-style unions with discriminants or similar
+mechanisms for identifying variants at runtime.
+
+## Other
+
+- should the [`#[projecting]`](#projecting-attribute) attribute have an associated field attribute
+  to mark the field that is projected onto?
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+## Enums
+
+Enums are difficult to support with the same framework as structs. The problem is that many
+containers don't provide sufficient guarantees to read the discriminant (for example raw pointers
+and `&mut MaybeUninit<T>`). However, for types that do provide sufficient guarantees, one could cook
+up a similar feature. Let's call them *enum projections*. They could work like this: projecting is
+done via a new kind of match operator:
+
+```rust
+enum MyEnum<F> {
+    A(i32, String),
+    B(#[pin] F),
+}
+type F = impl Future;
+let x: Pin<&mut MyEnum<F>>;
+match_proj x {
+    MyEnum::A(n, s) => {
+        let _: &mut i32 = n;
+        let _: &mut String = s;
+    }
+    MyEnum::B(fut) => {
+        let _: Pin<&mut F> = fut;
+    }
+}
+```
+
+I got this idea from reading the [Pinned places](https://without.boats/blog/pinned-places/) blog
+post from boats. There, enum projections for pinned references (i.e. just pin projections) are
+discussed.
+
+Here `match_proj` would need to be a new keyword. I dislike the name and syntax, but haven't come up
+with something better.
+
+A similar issue comes up in the design of [deref patterns]. Since the types `Pin` and `MyEnum` are
+distinct, they can be used to differentiate the kind of `match` the user wants to make. Thus making
+it possible to only have the `match` operator and not a separate `match_proj` operator.
+
+[deref patterns]: https://hackmd.io/4qDDMcvyQ-GDB089IPcHGg
+
+## Arrays
+
+Arrays can be thought of structs/tuples where each index is a field. Supporting them would simply
+follow tuples. They might need additional syntax or just use the tuple syntax.
+
+## Unions
+
+Since field access for unions is `unsafe`, projection would also have to be `unsafe`. Since unions
+are rarely used directly, this probably isn't important.
+
+## More Stdlib Additions
+
+Types that might be good candidates for [`#[projecting]`](#projecting-attribute):
+
+- `ManuallyDrop<T>`
+
+### `ArcRef<T>` for Stdlib
```

Comment:
FYI there's a crate for this: https://docs.rs/pared/latest/pared/sync/struct.Parc.html